/* Proto definition for final inference outputs. */

syntax = "proto3";

package tracking_pipeline.inference;

import "common.proto";
import "version.proto";


message AppearanceFeatures {
    repeated Appearance features = 1;
    uint32 dim = 2;
}

message Appearance {
    repeated float feats = 1;
}

enum DetectionPostureType {
    DETECTION_POSTURE_UNKNOWN = 0;
    DETECTION_POSTURE_STANDING = 1;
    DETECTION_POSTURE_SEATED = 2;
}

enum DetectionViewType {
    DETECTION_VIEW_UNKNOWN = 0;
    DETECTION_VIEW_FRONT = 1;
    DETECTION_VIEW_REAR = 2;
    DETECTION_VIEW_SIDE = 3;
}

message Track {
    string                      pid = 1;     // assigned pid
    repeated SingleViewTrack    single_view_tracks = 2; // single view tracks
    string                      face_feature_version = 3;
    string                      appearance_model_url = 4;
    string                      tracking_version = 5;
    ModelVersions               model_versions = 6;
    ModuleVersions              module_versions = 7;
}

// added by rlweng
message StaffRecognitionResult {
    // default value "0" means not-found/no-match from staff database.
    int32 staff_recog_id = 1;

    // corresponding face_id.
    int32 face_id = 2;

    float confidence = 3;
}

message SingleViewTrack {
    // basic information
    string              track_id = 1;       // original sv track id
    string              mv_linked_id = 2;   // mv track id
    string              video_name = 3;     // video name in case of offline processing (mostly for viz or pb matching)
    string              channel = 4;        // camera name
    repeated Detection  detections = 5;     // all detections
    bool                invalid = 6;        // reserved
    // attributes
    bytes               box_patches = 7;
    bytes               face_patches = 8;
    float               face_quality = 9;
    AppearanceFeatures  body_feature = 10;
    AppearanceFeatures  face_feature = 11;
    bytes               head_patches = 12;
    AppearanceFeatures  head_features = 13;
    AppearanceFeatures  mask_probs = 14;
    Box2D               best_face_box = 15;
    Box2D               origin_best_face_box = 16;
    BodyAttribute       body_attrs = 17; // added by wgchoi
    float               track_body_quality = 18; // added by rlweng
    StaffRecognitionResult staff_result = 19; // added by rlweng
}

message Detection {
    // basic information
    uint32  frame_index = 1; // video frame index
    float   human_detection_score = 2;
    int32   human_box_x = 3;
    int32   human_box_y = 4;
    int32   human_box_height = 5;
    int32   human_box_width = 6;
    uint64  det_time_msec = 7; // detection time in msec
    Point2D map_pos = 8; // localization in map
    int32   floor = 9;  // reserved

    // attributes
    DetectionPostureType    detection_posture_type = 10;
    repeated KeyPoint       key_points = 11;
    DetectionViewType       detection_view_type = 12;
    float head_angle = 13;
    float body_quality = 14; // added by rlweng
}
